### PR TITLE
drakrun: usermode profile generation using Injector

### LINKED ISSUE #472: Usermode profile generation: extract DLLs using injector -m readfile instead of locally accessing guest's filesystem
Right now during `draksetup postinstall`, we mount the filesystem using `qemu-nbd`, in order to extract system DLLs out of it. This could be replaced by spawning a toy VM (in a similar fashion which `drak-playground` works) and fetching DLLs using injector's `-m readfile` method.

related #463

### PR DESCRIPTION
**Pull request by manorit2001**
fix #472

---
This change is [Reviewable](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/486)

### CODE REVIEW

**Review by icedevml — COMMENTED** [review]
Manual test 4d1967273ed4c09e0273ffb25c1fe5a55d12576c

Looks like a good job in general. We need to test this one more thoroughly, as this is generally touching the critical functionality of `draksetup postinstall`.

Btw. there was some exception during my test:
```
[2021-03-14 13:36:52,618][DEBUG] Parsing PDB into JSON profile...
Traceback (most recent call last):
  File "/opt/venvs/drakrun/lib/python3.7/site-packages/drakrun/draksetup.py", line 410, in create_rekall_profiles
    profile = make_pdb_profile(tmp)
  File "/opt/venvs/drakrun/lib/python3.7/site-packages/drakrun/drakpdb.py", line 241, in make_pdb_profile
    pdb = pdbparse.parse(filepath)
  File "/opt/venvs/drakrun/lib/python3.7/site-packages/pdbparse/__init__.py", line 554, in parse
    return PDB7(f, fast_load)
  File "/opt/venvs/drakrun/lib/python3.7/site-packages/pdbparse/__init__.py", line 521, in __init__
    self.read_root(self.root_stream)
  File "/opt/venvs/drakrun/lib/python3.7/site-packages/pdbparse/__init__.py", line 467, in read_root
    parent = self))
  File "/opt/venvs/drakrun/lib/python3.7/site-packages/pdbparse/__init__.py", line 154, in __init__
    self.load()
  File "/opt/venvs/drakrun/lib/python3.7/site-packages/pdbparse/__init__.py", line 262, in load
    tpis = tpi.parse_stream(self.stream_file, unnamed_hack, elim_fwdrefs)
  File "/opt/venvs/drakrun/lib/python3.7/site-packages/pdbparse/tpi.py", line 1190, in parse_stream
    types[i].substructs = ListContainer([resolve_typerefs(t, types, min) for t in types[i].substructs])
  File "/opt/venvs/drakrun/lib/python3.7/site-packages/pdbparse/tpi.py", line 1190, in <listcomp>
    types[i].substructs = ListContainer([resolve_typerefs(t, types, min) for t in types[i].substructs])
  File "/opt/venvs/drakrun/lib/python3.7/site-packages/pdbparse/tpi.py", line 1105, in resolve_typerefs
    for attr in type_refs[leaf.leaf_type]:                                                                                                                                                   KeyError: 62195                                                                                                                                                                              [2021-03-14 13:37:01,664][WARNING] Unexpected exception while creating rekall profile for Windows/System32/ole32.dll, skipping...
```
but then the sandbox was setup succesfully and it worked.

I haven't seen such exception before, its worth to investigate whether this is related to this PR or just some unrelated bug.

**Comment by manorit2001** [comment]
> Manual test [4d19672](https://github.com/CERT-Polska/drakvuf-sandbox/commit/4d1967273ed4c09e0273ffb25c1fe5a55d12576c)
> 
> Looks like a good job in general. We need to test this one more thoroughly, as this is generally touching the critical functionality of `draksetup postinstall`.
> 
> Btw. there was some exception during my test:
> 
> ```
> [2021-03-14 13:36:52,618][DEBUG] Parsing PDB into JSON profile...
> Traceback (most recent call last):
>   File "/opt/venvs/drakrun/lib/python3.7/site-packages/drakrun/draksetup.py", line 410, in create_rekall_profiles
>     profile = make_pdb_profile(tmp)
>   File "/opt/venvs/drakrun/lib/python3.7/site-packages/drakrun/drakpdb.py", line 241, in make_pdb_profile
>     pdb = pdbparse.parse(filepath)
>   File "/opt/venvs/drakrun/lib/python3.7/site-packages/pdbparse/__init__.py", line 554, in parse
>     return PDB7(f, fast_load)
>   File "/opt/venvs/drakrun/lib/python3.7/site-packages/pdbparse/__init__.py", line 521, in __init__
>     self.read_root(self.root_stream)
>   File "/opt/venvs/drakrun/lib/python3.7/site-packages/pdbparse/__init__.py", line 467, in read_root
>     parent = self))
>   File "/opt/venvs/drakrun/lib/python3.7/site-packages/pdbparse/__init__.py", line 154, in __init__
>     self.load()
>   File "/opt/venvs/drakrun/lib/python3.7/site-packages/pdbparse/__init__.py", line 262, in load
>     tpis = tpi.parse_stream(self.stream_file, unnamed_hack, elim_fwdrefs)
>   File "/opt/venvs/drakrun/lib/python3.7/site-packages/pdbparse/tpi.py", line 1190, in parse_stream
>     types[i].substructs = ListContainer([resolve_typerefs(t, types, min) for t in types[i].substructs])
>   File "/opt/venvs/drakrun/lib/python3.7/site-packages/pdbparse/tpi.py", line 1190, in <listcomp>
>     types[i].substructs = ListContainer([resolve_typerefs(t, types, min) for t in types[i].substructs])
>   File "/opt/venvs/drakrun/lib/python3.7/site-packages/pdbparse/tpi.py", line 1105, in resolve_typerefs
>     for attr in type_refs[leaf.leaf_type]:                                                                                                                                                   KeyError: 62195                                                                                                                                                                              [2021-03-14 13:37:01,664][WARNING] Unexpected exception while creating rekall profile for Windows/System32/ole32.dll, skipping...
> ```
> 
> but then the sandbox was setup succesfully and it worked.
> 
> I haven't seen such exception before, its worth to investigate whether this is related to this PR or just some unrelated bug.

Yeah, this was a traceback actually, It is just being printed for debugging purpose if ever needed

File: drakrun/drakrun/draksetup.py
Comment (manorit2001) [inline drakrun/drakrun/draksetup.py:419]: @icedevml this line is responsible for the traceback you mentioned

Comment (manorit2001) [reply drakrun/drakrun/draksetup.py:419]: I can add a `[DEBUG]` tag before these debug tracebacks to avoid confusions
```
[2021-03-14 19:45:46,045][DEBUG] Traceback (most recent call last):
  File "/home/user/drakvuf-sandbox/drakrun/drakrun/draksetup.py", line 410, in create_rekall_profiles                                                                                        
    profile = make_pdb_profile(tmp)
  File "/home/user/drakvuf-sandbox/drakrun/drakrun/drakpdb.py", line 241, in make_pdb_profile
    pdb = pdbparse.parse(filepath)
  File "/home/user/drakvuf-sandbox/drakrun/env/lib/python3.7/site-packages/pdbparse/__init__.py", line 554, in parse                                                                         
    return PDB7(f, fast_load)
  File "/home/user/drakvuf-sandbox/drakrun/env/lib/python3.7/site-packages/pdbparse/__init__.py", line 521, in __init__                                                                      
    self.read_root(self.root_stream)
  File "/home/user/drakvuf-sandbox/drakrun/env/lib/python3.7/site-packages/pdbparse/__init__.py", line 467, in read_root                                                                     
    parent = self))
  File "/home/user/drakvuf-sandbox/drakrun/env/lib/python3.7/site-packages/pdbparse/__init__.py", line 154, in __init__                                                                      
    self.load()
  File "/home/user/drakvuf-sandbox/drakrun/env/lib/python3.7/site-packages/pdbparse/__init__.py", line 262, in load                                                                          
    tpis = tpi.parse_stream(self.stream_file, unnamed_hack, elim_fwdrefs)
  File "/home/user/drakvuf-sandbox/drakrun/env/lib/python3.7/site-packages/pdbparse/tpi.py", line 1190, in parse_stream                                                                      
    types[i].substructs = ListContainer([resolve_typerefs(t, types, min) for t in types[i].substructs])                                                                                      
  File "/home/user/drakvuf-sandbox/drakrun/env/lib/python3.7/site-packages/pdbparse/tpi.py", line 1190, in <listcomp>                                                                        
    types[i].substructs = ListContainer([resolve_typerefs(t, types, min) for t in types[i].substructs])                                                                                      
  File "/home/user/drakvuf-sandbox/drakrun/env/lib/python3.7/site-packages/pdbparse/tpi.py", line 1105, in resolve_typerefs                                                                  
    for attr in type_refs[leaf.leaf_type]:
KeyError: 62195

```

File: drakrun/drakrun/draksetup.py
Comment (manorit2001) [inline drakrun/drakrun/draksetup.py:420]: I can add a `[DEBUG]` tag before these debug tracebacks and show the warning above the traceback to avoid the confusions if it seems better
```suggestion
            logging.warning(f"Unexpected exception while creating rekall profile for {file.path}, skipping...")
            logging.debug(traceback.format_exc())
```

Giving the output like this,
```
[2021-03-14 19:45:32,409][DEBUG] Parsing PDB into JSON profile...   
[2021-03-14 19:45:46,045][WARNING] Unexpected exception while creating rekall profile for Windows/SysWOW64/ole32.dll, skipping...                                                                                                                                  
[2021-03-14 19:45:46,045][DEBUG] Traceback (most recent call last):                                                                                                                           
  File "/home/user/drakvuf-sandbox/drakrun/drakrun/draksetup.py", line 410, in create_rekall_profiles                                                                                         
    profile = make_pdb_profile(tmp)                                                                                                                                                           
  File "/home/user/drakvuf-sandbox/drakrun/drakrun/drakpdb.py", line 241, in make_pdb_profile                                                                                                 
    pdb = pdbparse.parse(filepath)                                                                                                                                                            
  File "/home/user/drakvuf-sandbox/drakrun/env/lib/python3.7/site-packages/pdbparse/__init__.py", line 554, in parse                                                                         
    return PDB7(f, fast_load)
  File "/home/user/drakvuf-sandbox/drakrun/env/lib/python3.7/site-packages/pdbparse/__init__.py", line 521, in __init__                                                                      
    self.read_root(self.root_stream)
  File "/home/user/drakvuf-sandbox/drakrun/env/lib/python3.7/site-packages/pdbparse/__init__.py", line 467, in read_root                                                                     
    parent = self))
  File "/home/user/drakvuf-sandbox/drakrun/env/lib/python3.7/site-packages/pdbparse/__init__.py", line 154, in __init__                                                                      
    self.load()
  File "/home/user/drakvuf-sandbox/drakrun/env/lib/python3.7/site-packages/pdbparse/__init__.py", line 262, in load                                                                          
    tpis = tpi.parse_stream(self.stream_file, unnamed_hack, elim_fwdrefs)
  File "/home/user/drakvuf-sandbox/drakrun/env/lib/python3.7/site-packages/pdbparse/tpi.py", line 1190, in parse_stream                                                                      
    types[i].substructs = ListContainer([resolve_typerefs(t, types, min) for t in types[i].substructs])                                                                                      
  File "/home/user/drakvuf-sandbox/drakrun/env/lib/python3.7/site-packages/pdbparse/tpi.py", line 1190, in <listcomp>                                                                        
    types[i].substructs = ListContainer([resolve_typerefs(t, types, min) for t in types[i].substructs])                                                                                      
  File "/home/user/drakvuf-sandbox/drakrun/env/lib/python3.7/site-packages/pdbparse/tpi.py", line 1105, in resolve_typerefs                                                                  
    for attr in type_refs[leaf.leaf_type]:
KeyError: 62195
                                                  
[2021-03-14 19:45:46,048][INFO] Fetching rekall profile for Windows/System32/combase.dll

```

Comment (icedevml) [reply drakrun/drakrun/draksetup.py:420]: So I understand that this error also appears for `combase.dll` in current version of Sandbox, but it's simply silenced?

Comment (manorit2001) [reply drakrun/drakrun/draksetup.py:420]: I don't really have track for what all DLLs this error occurs but yes, This was silenced before when the exception was handled

Comment (manorit2001) [reply drakrun/drakrun/draksetup.py:420]: I modified the code a bit locally to capture the DLLs categorically which cause errors, here are the results
- `Error DLLs` show the traceback which you had seen

Other names should be intuitive ig
```
[2021-03-15 02:07:00,487][DEBUG] Error DLLs: ['Windows/System32/ole32.dll', 'Windows/SysWOW64/ole32.dll']
[2021-03-15 02:07:00,490][DEBUG] runtime error DLLs: []
[2021-03-15 02:07:00,490][DEBUG] failed_to_copy DLLs: ['Windows/System32/combase.dll']

```

Comment (icedevml) [reply drakrun/drakrun/draksetup.py:420]: Cool!

File: drakrun/drakrun/draksetup.py
Comment (chivay) [inline drakrun/drakrun/draksetup.py:570]: I think it woud be a good idea to create a VM object in this function to handle saving and destroying the VM.

File: drakrun/drakrun/draksetup.py
Comment (chivay) [inline drakrun/drakrun/draksetup.py:557]: How about adding a new method to `VirtualMachine` class? For example:
`def save(self, file_path: str, destroy=True)`

File: drakrun/drakrun/draksetup.py
Comment (chivay) [inline drakrun/drakrun/draksetup.py:403]: I'm not sure about using join to concatenate Windows paths on Linux systems.
IMO in this case we should use `PureWindowsPath` from `pathlib`.

**Review by chivay — COMMENTED** [review]
Only some minor issues. Overall a very nice and welcome change!

File: drakrun/drakrun/draksetup.py
Comment (manorit2001) [reply drakrun/drakrun/draksetup.py:570]: yeah, I was thinking about it while working on VM unit tests. I have more refactoring ideas also which I wanted to discuss regarding the `vm.py`. That PR will address `vm.py` related stuff. It will also be easy to track these refactoring changes that way ig. Will do this in that PR if it seems fine?

File: drakrun/drakrun/draksetup.py
Comment (manorit2001) [reply drakrun/drakrun/draksetup.py:403]: Yeah, I was confused about this and I also had added a comment in (344c5cb391b2546d5bae) which I removed recently in (https://github.com/CERT-Polska/drakvuf-sandbox/pull/486/commits/4d1967273ed4c09e0273ffb25c1fe5a55d12576c) after doing a manual test and not getting any issues. Though, I will change it. Thanks

File: drakrun/drakrun/draksetup.py
Comment (manorit2001) [reply drakrun/drakrun/draksetup.py:557]: Same with this change also(regarding different PR) if it seems fine

File: drakrun/drakrun/draksetup.py
Comment (chivay) [reply drakrun/drakrun/draksetup.py:403]: :+1:

File: drakrun/drakrun/draksetup.py
Comment (chivay) [reply drakrun/drakrun/draksetup.py:570]: :+1:

File: drakrun/drakrun/draksetup.py
Comment (chivay) [reply drakrun/drakrun/draksetup.py:557]: :+1:

File: drakrun/drakrun/draksetup.py
Comment (chivay) [inline drakrun/drakrun/draksetup.py:424]: We can use `safe_delete`

File: drakrun/drakrun/draksetup.py
Comment (chivay) [inline drakrun/drakrun/draksetup.py:427]: Instead of this, I think it would be simpler to initialize `tmp` with `None` and then check `if tmp is not None and os.path.exists(..)`

File: drakrun/drakrun/draksetup.py
Comment (manorit2001) [reply drakrun/drakrun/draksetup.py:424]: Right!

File: drakrun/drakrun/draksetup.py
Comment (chivay) [inline drakrun/drakrun/draksetup.py:396]: One small nitpick: install_info now becomes redundant.
I would pass `Injector` directly into this function

**Review by chivay — APPROVED** [review]

File: drakrun/drakrun/draksetup.py
Comment (manorit2001) [reply drakrun/drakrun/draksetup.py:396]: Sure, will change it

**Comment by icedevml** [comment]
I need to review this but this is going to take few days.

**Comment by manorit2001** [comment]
> I need to review this but this is going to take few days.

Sure, no issues

**Review by icedevml — COMMENTED** [review]
> Reviewed 1 of 2 files at r1, 1 of 1 files at r5, 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/486)* status: all files reviewed, 8 unresolved discussions (waiting on @chivay and @manorit2001)

**Review by chivay — APPROVED** [review]
> Reviewed 1 of 2 files at r1, 1 of 1 files at r5, 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/486)* status: all files reviewed, 2 unresolved discussions (waiting on @manorit2001)

**Review by manorit2001 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/486)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @manorit2001)

___
*[drakrun/drakrun/draksetup.py, line 403 at r2](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/486#-MVqBN4PZb0M9G2DpckD-r2-403:-MVvWnet9oy2sGwAhNc0:b-896fix) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/2627a564a8404145315946a258d4adb7f00b4f05/drakrun/drakrun/draksetup.py#L403)):*
<details><summary><i>Previously, chivay (Hubert Jasudowicz) wrote…</i></summary><blockquote>

:+1: 
</blockquote></details>

Done.
___
*[drakrun/drakrun/draksetup.py, line 427 at r3](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/486#-MVqBN4PZb0M9G2DpckD-r3-427:-MVvX0Kp7vWMS7xbgFb8:b-896fix) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/7bd1a4d2576176678e5c5d59892080e651c093d0/drakrun/drakrun/draksetup.py#L427)):*
<details><summary><i>Previously, chivay (Hubert Jasudowicz) wrote…</i></summary><blockquote>

Instead of this, I think it would be simpler to initialize `tmp` with `None` and then check `if tmp is not None and os.path.exists(..)`
</blockquote></details>

Done.
___
*[drakrun/drakrun/draksetup.py, line 396 at r4](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/486#-MVqBN4PZb0M9G2DpckD-r4-396:-MVvWlkN7JcrkT5p_wat:b-896fix) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/d3cff1f6fd243ab2d8869719699cfe479afb7fab/drakrun/drakrun/draksetup.py#L396)):*
<details><summary><i>Previously, manorit2001 (Manorit Chawdhry) wrote…</i></summary><blockquote>

Sure, will change it
</blockquote></details>

Done.

**Comment by icedevml** [comment]
I've made a PR to upstream to improve logging&error handling in injector's `-m readfile` and `-m writefile`:
https://github.com/tklengyel/drakvuf/pull/1171

I think this one is going to be useful here. Let's wait until this upstream PR is merged and then I will give you some further instructions.

I've tested this and it generally looks very cool, I would like to ask you to make one (rather small) change, but first we need to wait for the above PR to be merged.

**Comment by manorit2001** [comment]
> I've made a PR to upstream to improve logging&error handling in injector's `-m readfile` and `-m writefile`:
> [tklengyel/drakvuf#1171](https://github.com/tklengyel/drakvuf/pull/1171)
> 
> I think this one is going to be useful here. Let's wait until this upstream PR is merged and then I will give you some further instructions.

Yeah, I agree. The errors and logging weren't good in that

> I've tested this and it generally looks very cool, I would like to ask you to make one (rather small) change, but first we need to wait for the above PR to be merged.

Thanks! Sure, will await your further instructions

**Comment by icedevml** [comment]
Ok, please bump DRAKVUF submodule to the latest master that includes merged PR 1171. With this PR, injector (when ran with `-o json`) should print a single line to stdout with a JSON object, that would describe the exact injection status.

Valid run (`ProcessName` should be understood here as `FileName`):
```
{"Plugin": "inject", "TimeStamp": "1615922048.468544", "Status": "Success", "ProcessName": "C:\\windows\\system32\\ntdll.dll", "Arguments": "", "InjectedPid": 0, "InjectedTid": 0}
```

file not found:
```
{"Plugin": "inject", "TimeStamp": "1615922052.607144", "Status": "Error", "ErrorCode": 2, "Error": "ERROR_FILE_NOT_FOUND"}
```

some other errors:
```
{"Plugin": "inject", "TimeStamp": "1615922057.550625", "Status": "Error", "ErrorCode": 6, "Error": "ERROR_INVALID_HANDLE"}
```

failed to initialize injector:
```
{"Plugin": "inject", "TimeStamp": "1615922110.075010", "Status": "InitFail"}
```

So we could run injector with `-o json`, then capture the stdout of injector (regardless of the exit code), then try to run `json.loads` on it, just to find out the exact reason why injector has failed.

If there is an empty output or it can not be parsed as JSON, `draksetup` should print JSON parse exception and fail.

I think `if status == "Error" and error == "ERROR_FILE_NOT_FOUND"` then this is tolerable, we could just log that some DLL was not found and continue.

If there is any other error condition, we should print all the information to the user (I mean the whole object `{"Plugin": "inject", ...}`) and draksetup should fail, because something went really wrong here.

**Comment by icedevml** [comment]
> [2021-03-17 21:24:22,281][DEBUG] stderr: DRAKVUF injector v0.8-git20201127091009+52db64e-1 Copyright (C) 2014-2020 Tamas K Lengyel

This is from some old injector (revision 2020-11-27), this might be one of the bugs that were fixed in the recent revisions.

**Comment by icedevml** [comment]
@manorit2001 The test on the CI might be failing for unrelated reason, I will look at it.

**Comment by icedevml** [comment]
@manorit2001 Ah, we've slightly diverged from the upstream DRAKVUF. Please bump DRAKVUF to this branch: https://github.com/CERT-Polska/drakvuf/tree/sandbox-20210317-mq1 instead of master. The CI will work then.

We will fix this divergence but later.

**Review by icedevml — CHANGES_REQUESTED** [review]

File: drakrun/drakrun/draksetup.py
Comment (manorit2001) [reply drakrun/drakrun/draksetup.py:577]: Ah, shit! I really didn't consider the race conditions. And, umm apparently `Applying suggestions on deleted lines is not supported` as per GH. My commit suggestion button is greyed out. I will write and commit this manually then

**Comment by manorit2001** [comment]
> > [2021-03-17 21:24:22,281][DEBUG] stderr: DRAKVUF injector v0.8-git20201127091009+52db64e-1 Copyright (C) 2014-2020 Tamas K Lengyel
> 
> This is from some old injector (revision 2020-11-27), this might be one of the bugs that were fixed in the recent revisions.

Oh, I need to update my bundle then ig

**Comment by manorit2001** [comment]
@icedevml could you please review again? Locally this is working fine for me with the new `injector`

**Comment by icedevml** [comment]
Sorry for the delay, if everything is OK then this will be merged soon.

**Review by icedevml — APPROVED** [review]

**Comment by icedevml** [comment]
@manorit2001 Huge thanks for this! Sorry for the delay, but as this PR is touching a crucial feature, we had to be really careful about it.

**Pull request merged**
